ssh-cipher/src/chacha20poly1305.rs

```diff
@@ -94,12 +94,8 @@ impl Cipher {
     /// Encrypt the provided `buffer` in-place, returning the Poly1305 authentication tag.
     #[inline]
     pub fn encrypt(mut self, associated_data: &[u8], buffer: &mut [u8]) -> Result<Tag> {
-        // TODO(tarcieri): support associated data (RustCrypto/SSH#279)
-        if !associated_data.is_empty() {
-            return Err(Error);
-        }
-
         self.cipher.apply_keystream(buffer);
+        self.mac.update_padded(associated_data);
```

Comment:
Ugh, well that isn't going to work. It needs to be unpadded, and unfortunately [`Poly1305::compute_unpadded`](https://docs.rs/poly1305/latest/poly1305/struct.Poly1305.html#method.compute_unpadded) only works on a contiguous slice.